domain/resource/state/resource.go

```diff
@@ -407,198 +421,543 @@ WHERE uuid = $resourceIdentity.uuid`,
+func (st *State) GetResourceType(
 	ctx context.Context,
-	config resource.SetUnitResourceArgs,
-) (resource.SetUnitResourceResult, error) {
+	resourceUUID coreresource.UUID,
+) (charmresource.Type, error) {
 	db, err := st.DB()
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return 0, errors.Capture(err)
 	}
 
-	// Prepare statement to check if the unit/resource is not already there.
-	unitResourceInput := unitResource{
-		ResourceUUID: config.ResourceUUID.String(),
-		UnitUUID:     config.UnitUUID.String(),
-		AddedAt:      st.clock.Now(),
+	resKind := resourceKind{
+		UUID: resourceUUID.String(),
 	}
-	checkUnitResourceQuery := `
-SELECT &unitResource.* FROM unit_resource 
-WHERE unit_resource.resource_uuid = $unitResource.resource_uuid 
-AND unit_resource.unit_uuid = $unitResource.unit_uuid`
-	checkUnitResourceStmt, err := st.Prepare(checkUnitResourceQuery, unitResourceInput)
+	getResourceType, err := st.Prepare(`
+SELECT &resourceKind.kind_name 
+FROM   v_resource
+WHERE  uuid = $resourceKind.uuid
+`, resKind)
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return 0, errors.Capture(err)
 	}
 
-	// Prepare statement to check that UnitUUID is valid UUID.
-	unitUUID := unitNameAndUUID{UnitUUID: config.UnitUUID}
-	checkValidUnitQuery := `
-SELECT &unitNameAndUUID.uuid 
-FROM unit 
-WHERE uuid = $unitNameAndUUID.uuid`
-	checkValidUnitStmt, err := st.Prepare(checkValidUnitQuery, unitUUID)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = tx.Query(ctx, getResourceType, resKind).Get(&resKind)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return resourceerrors.ResourceNotFound
+		}
+		return errors.Capture(err)
+	})
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return 0, errors.Capture(err)
 	}
 
-	// Prepare statement to check that resourceID is valid UUID.
-	resourceUUID := resourceIdentity{UUID: config.ResourceUUID.String()}
-	checkValidResourceQuery := `
-SELECT &resourceIdentity.uuid
-FROM resource
-WHERE uuid = $resourceIdentity.uuid`
-	checkValidResourceStmt, err := st.Prepare(checkValidResourceQuery, resourceUUID)
+	kind, err := charmresource.ParseType(resKind.Name)
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return 0, errors.Errorf("parsing resource kind: %w", err)
+	}
+	return kind, err
+}
+
+// insertFileResource checks that the storage ID corresponds to stored object
+// store metadata and then records that the resource is stored at the provided
+// storage ID.
+func (st *State) insertFileResource(
+	ctx context.Context,
+	tx *sqlair.TX,
+	resourceUUID coreresource.UUID,
+	storageID coreresourcestore.ID,
+) error {
+	// Get the object store UUID of the stored resource blob.
+	if !storageID.IsObjectStoreUUID() {
+		return errors.Errorf("cannot insert file resource that is not stored in the object store")
+	}
+	uuid, err := storageID.ObjectStoreUUID()
```

Comment:
**suggestion:** I don't think it is useful to have two check (the check that is a objectStoreUUID and a second check. I think we can skip the `IsObjectStoreUUID` and annotate the error returned by `storageID.ObjectStoreUUID`  instead.

Reply:
Good point, will do that.